File: apps/android/src/main/java/tv/every/tilawah/android/storage/InMemoryHistoryStore.kt
package tv.every.tilawah.android.storage

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import java.time.Instant

/**
 * In-process [HistoryStore] for unit tests and Compose previews.
 * Mirrors the production [DataStoreHistoryStore] semantics (50-attempt
 * cap, best-of replacement).
 */
class InMemoryHistoryStore : HistoryStore {

    private val _last = MutableStateFlow<LastPracticed?>(null)
    private val _best = MutableStateFlow<Map<String, BestScoreEntry>>(emptyMap())
    private val _attempts = MutableStateFlow<List<Attempt>>(emptyList())

    override fun lastPracticed(): Flow<LastPracticed?> = _last.asStateFlow()

    override suspend fun setLastPracticed(entry: LastPracticed) {
        _last.value = entry
    }

    override fun bestScore(surahId: String, ayahNumber: Int): Flow<BestScoreEntry?> =
        _best.map { it[bestScoreKey(surahId, ayahNumber)] }

    override fun bestScoreForSurah(surahId: String): Flow<Double?> =
        _best.map { all ->
            val prefix = "$surahId:"
            all.entries
                .filter { it.key.startsWith(prefix) }
                .maxOfOrNull { it.value.score }
        }

    override suspend fun recordBestScore(
        surahId: String,
        ayahNumber: Int,
        score: Double,
        achievedAt: Instant,
    ) {
        val key = bestScoreKey(surahId, ayahNumber)
        val existing = _best.value[key]
        if (existing == null || score > existing.score) {
            _best.value = _best.value + (key to BestScoreEntry(score, achievedAt))
        }
    }

    override fun recentAttempts(limit: Int): Flow<List<Attempt>> =
        _attempts.map { it.take(limit) }

    override suspend fun recordAttempt(attempt: Attempt) {
        _attempts.value = (listOf(attempt) + _attempts.value)
            .take(HistoryStoreConstants.HISTORY_LIMIT)
    }
}

File: apps/ios/Sources/QuranRecitationApp/Storage/InMemoryHistoryStore.swift
import Foundation

/// In-memory `HistoryStore` for SwiftUI Previews and unit tests. State
/// is held in plain dictionaries; nothing is persisted.
final class InMemoryHistoryStore: HistoryStore {
  private var lastPracticedEntry: LastPracticed?
  private var bestScores: [String: BestScoreEntry] = [:]
  private var attempts: [Attempt] = []

  init(
    lastPracticed: LastPracticed? = nil,
    bestScores: [String: BestScoreEntry] = [:],
    attempts: [Attempt] = []
  ) {
    self.lastPracticedEntry = lastPracticed
    self.bestScores = bestScores
    self.attempts = attempts
  }

  func lastPracticed() -> LastPracticed? {
    lastPracticedEntry
  }

  func setLastPracticed(_ entry: LastPracticed) {
    lastPracticedEntry = entry
  }

  func bestScore(surahId: String, ayahNumber: Int) -> BestScoreEntry? {
    bestScores["\(surahId):\(ayahNumber)"]
  }

  func bestScore(forSurah surahId: String) -> Double? {
    let prefix = "\(surahId):"
    return bestScores
      .filter { $0.key.hasPrefix(prefix) }
      .map(\.value.score)
      .max()
  }

  func recordBestScore(surahId: String, ayahNumber: Int, score: Double, achievedAt: Date) {
    let key = "\(surahId):\(ayahNumber)"
    if let existing = bestScores[key], existing.score >= score { return }
    bestScores[key] = BestScoreEntry(score: score, achievedAt: achievedAt)
  }

  func recentAttempts(limit: Int) -> [Attempt] {
    Array(attempts.prefix(limit))
  }

  func recordAttempt(_ attempt: Attempt) {
    attempts.insert(attempt, at: 0)
    if attempts.count > HistoryStoreConstants.historyLimit {
      attempts = Array(attempts.prefix(HistoryStoreConstants.historyLimit))
    }
  }

  func clear() {
    lastPracticedEntry = nil
    bestScores = [:]
    attempts = []
  }

  @MainActor
  func refreshFromRemote() async {
    // No remote; in-memory state is the source of truth.
  }
}
